crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -1256,6 +1265,162 @@ impl IndexerReader {
+        } else {
+            None
+        };
+
+        let effects = if options.show_effects {
+            let effects = effects_option
+                .ok_or_else(|| IndexerError::GenericError("Effects should be available".into()))?;
+            Some(SuiTransactionBlockEffects::try_from(effects)?)
+        } else {
+            None
+        };
+
+        let raw_transaction = if options.show_raw_input {
+            stored_transaction.raw_transaction
+        } else {
+            Vec::new()
+        };
+
+        let events = if options.show_events {
+            let events = {
+                stored_transaction
+                        .events
+                        .into_iter()
+                        .map(|event| match event {
+                            Some(event) => {
+                                let event: Event = bcs::from_bytes(&event).map_err(|e| {
+                                    IndexerError::PersistentStorageDataCorruptionError(format!(
+                                        "Can't convert event bytes into Event. tx_digest={:?} Error: {e}",
+                                        tx_digest
+                                    ))
+                                })?;
+                                Ok(event)
+                            }
+                            None => Err(IndexerError::PersistentStorageDataCorruptionError(format!(
+                                "Event should not be null, tx_digest={:?}",
+                                tx_digest
+                            ))),
+                        })
+                        .collect::<Result<Vec<Event>, IndexerError>>()?
+            };
+            let timestamp = stored_transaction.timestamp_ms as u64;
+            let tx_events = TransactionEvents { data: events };
+
+            tx_events_to_sui_tx_events(tx_events, package_resolver, tx_digest, timestamp).await?
+        } else {
+            None
+        };
+
+        let balance_changes = if options.show_balance_changes {
+            let balance_changes = {
+                stored_transaction.balance_changes.into_iter().map(|balance_change| {
+                        match balance_change {
+                            Some(balance_change) => {
+                                let balance_change: BalanceChange = bcs::from_bytes(&balance_change)
+                                    .map_err(|e| IndexerError::PersistentStorageDataCorruptionError(
+                                        format!("Can't convert balance_change bytes into BalanceChange. tx_digest={:?} Error: {e}", tx_digest)
+                                    ))?;
+                                Ok(balance_change)
+                            }
+                            None => Err(IndexerError::PersistentStorageDataCorruptionError(format!("object_change should not be null, tx_digest={:?}", tx_digest))),
+                        }
+                    }).collect::<Result<Vec<BalanceChange>, IndexerError>>()?
```

Comment:
Same here also...and i wonder if it would be better to store balance changes as a json payload vs a bcs one...

Reply:
same, agreed with the vector part, what's the motivation of changing bcs to json? I was under the impression that bcs bytes is generally more compact.

Reply:
> what's the motivation of changing bcs to json? I was under the impression that bcs bytes is generally more compact.

its less about the compactness and more about the fact that its not a "core" protocol type. json is more evolvable than bcs is as well...

Reply:
yeah that sounds fair to me, graphql is using the bcs bytes too, let's make a post to ask around? I am happy to make the code changes when we have aligntment.